cli/command/system/dial_stdio.go

```diff
@@ -0,0 +1,125 @@
+package system
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type dialStdioOptions struct {
+	// equivalent of `socat -t` and `nc -q`
+	timeout time.Duration
+}
+
+// newDialStdioCommand creates a new cobra.Command for `docker system dial-stdio`
+func newDialStdioCommand(dockerCli command.Cli) *cobra.Command {
+	var opts dialStdioOptions
+
+	cmd := &cobra.Command{
+		Use:    "dial-stdio [OPTIONS]",
+		Short:  "Proxy the stdio stream to the daemon connection. Should not be invoked manually.",
+		Args:   cli.NoArgs,
+		Hidden: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runDialStdio(dockerCli, opts)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.DurationVarP(&opts.timeout, "timeout", "t", 500*time.Millisecond, "After EOF from one of inputs, wait for this duration before exit")
+	return cmd
+}
+
+func runDialStdio(dockerCli command.Cli, opts dialStdioOptions) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	conn, err := dockerCli.Client().DialRaw(ctx)
```

Comment:
I think this would be slightly better as a `Dialer() func(ctx) (net.Conn, error)` . Otherwise, it seems like a method that calls to the remote side. We could also solve this by breaking up `NewClientWithOpts` so that we always get access to dialer in cli before creating the client and store the dialer so we can reuse it.